packages/graphql/tests/e2e/federation/apollo-federation-subgraph-compatibility/apollo-federation-subgraph-compatibility.e2e.test.ts

```diff
@@ -0,0 +1,675 @@
+            );
+        });
+
+        test("resolves multiple @key directives on Product", async () => {
+            const entitiesQuery = await productsRequest({
+                query: `#graphql
+          query ($representations: [_Any!]!) {
+            _entities(representations: $representations) {
+              ...on Product { id sku }
+            }
+          }
+        `,
+                variables: {
+                    representations: [
+                        {
+                            __typename: "Product",
+                            id: "apollo-federation",
+                        },
+                        {
+                            __typename: "Product",
+                            sku: "federation",
+                            package: "@apollo/federation",
+                        },
+                        {
+                            __typename: "Product",
+                            sku: "studio",
+                            variation: { id: "platform" },
+                        },
+                    ],
+                },
+            });
+
+            expect(entitiesQuery).not.toHaveProperty("errors");
+            expect(entitiesQuery).toMatchObject({
+                data: {
+                    _entities: [
+                        {
+                            id: "apollo-federation",
+                            sku: "federation",
+                        },
+                        {
+                            id: "apollo-federation",
+                            sku: "federation",
+                        },
+                        {
+                            id: "apollo-studio",
+                            sku: "studio",
+                        },
+                    ],
+                },
+            });
+        });
+    });
+
+    test("@link", async () => {
+        const response = await productsRequest({
+            query: "query { _service { sdl } }",
+        });
+
+        expect(response.data).toMatchObject({
+            _service: {
+                sdl: expect.stringContaining("@link"),
+            },
+        });
+
+        const { sdl } = response.data._service;
+        const normalizedSDL = stripIgnoredCharacters(sdl);
+
+        const linksRegex = /@link\(([\s\S]+?)\)/g;
+        // should have @link imports
+        expect(linksRegex.test(normalizedSDL)).toBe(true);
+
+        let fedLinkCount = 0;
+        normalizedSDL.match(linksRegex)?.forEach((element) => {
+            const urlRegex = /url:(".+?")/;
+            // skip definitions
+            if (urlRegex.test(element)) {
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                const linkUrl = JSON.parse(element.match(urlRegex)![1]);
+                const linkUrlSpecVersionRegex = /https:\/\/specs.apollo.dev\/federation\/v(.+)/;
+                // only verify federation spec @links
```

Comment:
I guess there are not many ways of testing this, but this complexity on a test could easily lead to flaky tests, consider refactoring this to simplify the logic a bit if possible

Reply:
This whole thing will disappear when we migrate over to using the GitHub Action from Apollo.